bin/update-pot-file-references.php

```diff
@@ -0,0 +1,160 @@
+			$references            = [];
+			$messages              = [];
+			$have_msgid            = false;
+			continue;
+		}
+
+		if ( 'msgid' == substr( $line, 0, 5 ) ) {
+			$have_msgid = true;
+		}
+
+		if ( $have_msgid ) {
+			$messages[] = $line;
+		} else {
+			$references[] = $line;
+		}
+	}
+
+	fclose( $fh );
+
+	$message               = implode( "\n", $messages );
+	$originals[ $message ] = $references;
+	return $originals;
+}
+
+/**
+ * Generates a map with the mapping for 'original source file' -> final transpiled/minified file in the 'dist' folder.
+ * Format example:
+ * [
+ *   'client/card-readers/settings/file-upload.js' => [
+ *     'build/index.js',
+ *     'build/tos.js',
+ *   ]
+ * ]
+ *
+ * @return array Mapping of source js files and the generated files that use them.
+ */
+function load_js_transpiling_source_maps(): array {
+	$mappings = [];
+	foreach ( glob( "build/*.js.map", GLOB_NOSORT ) as $filename ) {
+		$file_content = file_get_contents( $filename );
+		if ( $file_content === false ) {
+			echo "[WARN] Unable to read file '". $filename . "'. Some translation strings might not have the correct references as a result.\n";
+			continue;
+		}
+		$file_json = json_decode( $file_content, true );
+		if ( $file_json === null ) {
+			echo "[WARN] Unable to parse JSON file: '". $filename . "'. Some translation strings might not have the correct references as a result.\n";
+			continue;
+		}
+
+		foreach ( $file_json[ 'sources' ] as $source ) {
+			$source = preg_replace( '%^webpack:///\./(client/.*)$%', '${1}', $source );
+			if ( 'webpack' !== substr( $source, 0, 7 ) ) {
+				$mappings[ $source ][] = $file_json[ 'file' ];
+			}
+		}
+	}
+
+	if ( empty( $mappings ) ) {
+		echo "[ERROR] Unable to load JS transpiling mappings from 'build/*.js.map' files. Make sure the JS assets compilation was successful.\n";
+		die();
+	}
+
+	return $mappings;
+}
+
+/**
+ * For each file reference to a javascript/typescript file (from the client folder) in the comments, it adds file
+ * references to the generated files (from the build folder) that use that particular javascript/typescript as source.
+ *
+ * @param array $js_mappings Mapping of source js files and the generated files that use them.
+ * @param array $translations POT translations (including references/comments).
+ * @return array Translation messages
+ */
+function add_transpiled_filepath_reference_to_comments( array $js_mappings, array $translations ): array {
+	foreach ( $translations as $message => $references ) {
+		// Check references for js/jsx/ts/tsx files
+		$dist_js_to_add = [];
+		foreach ( $references as $ref ) {
+			if ( preg_match( '%^#: (.+\.(js|jsx|ts|tsx)):\d+$%', $ref, $m ) ) {
+				if ( ! array_key_exists( $m[1], $js_mappings ) ) {
+					// The file $m[1] is not used in any of the generated client JS files. Skip it.
+					continue;
+				}
+
+				foreach ( $js_mappings[ $m[1] ] as $mapping ) {
+					$dist_js_to_add[] = '#: build/' . $mapping . ':1';
```

Comment:
This won't work if build files are formatted or not inlined right? In any case, is it important to reference the exact line number the string is referenced?

Reply:
It only adds a `file reference comment` so we can generate JSON files for each entrypoint (transpiled javascripts in the `build` folder), the line number 1 is there for consistency (the content of the file is never parsed, only the filename is used), also the actual `occurrences` of the strings are still there to help translators:
```
#: build/settings.js:1
#: client/additional-methods-setup/methods-selector/add-payment-methods-task.js:391
#: client/connect-account-page/modal/index.js:82
#: client/multi-currency-setup/tasks/store-settings-task/index.js:147
#: client/payment-methods/activation-modal.js:47
msgid "Continue"
msgstr ""
```
> More context on the original WC Payments issue in the slack thread: p1638277640394300-slack-CGGCLBN58